src/Mobs/PathFinder.cpp

```diff
@@ -0,0 +1,226 @@
+			break;
+		}
+
+		case ePathFinderStatus::PATH_NOT_FOUND:
+		{
+			return ePathFinderStatus::PATH_NOT_FOUND;
+		}
+		case ePathFinderStatus::CALCULATING:
+		{
+			m_TicksSinceLastPathReset = 0;  // Start ticking only after a path is found.
+			return ePathFinderStatus::CALCULATING;
+		}
+		case ePathFinderStatus::PATH_FOUND:
+		{
+			if (m_NoMoreWayPoints || (--m_GiveUpCounter == 0))
+			{
+				ResetPathFinding();  // Try to calculate a path again.
+				return ePathFinderStatus::CALCULATING;
+			}
+			else if (!m_Path->IsLastPoint())  // Have we arrived at the next cell, as denoted by m_NextWayPointPosition?
+			{
+				if ((m_Path->IsFirstPoint() || (m_NextWayPointPosition - m_Source).SqrLength() < 0.25))
+				{
+					m_NextWayPointPosition = m_Path->GetNextPoint();
+					a_OutputWaypoint = m_NextWayPointPosition;
+					m_GiveUpCounter = 40;  // Give up after 40 ticks (2 seconds) if failed to reach m_NextWayPointPosition.
+				}
+			}
+			else
+			{
+				m_NoMoreWayPoints = true;
+			}
+
+			m_IsFollowingPath = true;
+			return ePathFinderStatus::PATH_FOUND;
+		}
+
+	default:  // Dummy to get rid of compiler warning.
+		return ePathFinderStatus::PATH_FOUND;
+		break;
+	}
+}
+
+
+void cPathFinder::ResetPathFinding(void)
+{
+	m_TicksSinceLastPathReset = 0;
+	m_IsFollowingPath = false;
+	m_NoMoreWayPoints = false;
+	if (m_Path != nullptr)
+	{
+		delete m_Path;
+		m_Path = nullptr;
+	}
+}
+
+
+bool cPathFinder::EnsureProperDestination(cChunk & a_Chunk)
+{
+	cChunk * Chunk = a_Chunk.GetNeighborChunk(FloorC(m_FinalDestination.x), FloorC(m_FinalDestination.z));
+	BLOCKTYPE BlockType;
+	NIBBLETYPE BlockMeta;
+
+	if ((Chunk == nullptr) || !Chunk->IsValid())
+	{
+		return false;
+	}
+
+	int RelX = FloorC(m_FinalDestination.x) - Chunk->GetPosX() * cChunkDef::Width;
+	int RelZ = FloorC(m_FinalDestination.z) - Chunk->GetPosZ() * cChunkDef::Width;
+
+	// If destination in the air, first try to go 1 block north, or east, or west.
+	// This fixes the player leaning issue.
+	// If that failed, we instead go down to the lowest air block.
+	Chunk->GetBlockTypeMeta(RelX, FloorC(m_FinalDestination.y) - 1, RelZ, BlockType, BlockMeta);
+	if (!cBlockInfo::IsSolid(BlockType))
+	{
+		bool InTheAir = true;
+		int x, z;
+		for (z = -1; z <= 1; ++z)
+		{
+			for (x = -1; x <= 1; ++x)
+			{
+				if ((x==0) && (z==0))
+				{
+					continue;
+				}
+				Chunk = a_Chunk.GetNeighborChunk(FloorC(m_FinalDestination.x+x), FloorC(m_FinalDestination.z+z));
+				if ((Chunk == nullptr) || !Chunk->IsValid())
+				{
+					return false;
+				}
+				RelX = FloorC(m_FinalDestination.x+x) - Chunk->GetPosX() * cChunkDef::Width;
+				RelZ = FloorC(m_FinalDestination.z+z) - Chunk->GetPosZ() * cChunkDef::Width;
+				Chunk->GetBlockTypeMeta(RelX, FloorC(m_FinalDestination.y) - 1, RelZ, BlockType, BlockMeta);
+				if (cBlockInfo::IsSolid(BlockType))
+				{
+					m_FinalDestination.x += x;
+					m_FinalDestination.z += z;
+					InTheAir = false;
+					goto breakBothLoops;
```

Comment:
I'd really prefer to live without them, if at all possible.

In this case, I'd consider replacing the two `for`-loops with a single one by listing all the coords it checks in a `static const` array of 2D coords and iterating over that array, using a regular `break` instead of the `goto`.